exporter/carbonexporter/metricdata_to_plaintext.go

```diff
@@ -0,0 +1,388 @@
+
+	// Constants used when converting from summary metrics to Carbon format.
+	summaryQuantileSuffix         = ".quantile"
+	summaryQuantileTagKey         = "quantile"
+	summaryQuantileTagBeforeValue = tagPrefix + summaryQuantileTagKey + tagKeyValueSeparator
+
+	// Suffix to be added to original metric name for a Carbon metric representing
+	// a count metric for either distribution or summary metrics.
+	countSuffix = ".count"
+
+	// Textual representation for positive infinity valid in Carbon, ie.:
+	// positive infinity as represented in Python.
+	infinityCarbonValue = "inf"
+)
+
+// metricDataToPlaintext converts internal metrics data to the Carbon plaintext
+// format as defined in https://graphite.readthedocs.io/en/latest/feeding-carbon.html#the-plaintext-protocol)
+// and https://graphite.readthedocs.io/en/latest/tags.html#carbon. See details
+// below.
+//
+// Each metric point becomes a single string with the following format:
+//
+// 	"<path> <value> <timestamp>"
+//
+// The <path> contains the metric name and its tags and has the following,
+// format:
+//
+// 	<metric_name>[;tag0;...;tagN]
+//
+// <metric_name> is the name of the metric and terminates either at the first ';'
+// or at the end of the path.
+//
+// <tag> is of the form "key=val", where key can contain any char except ";!^=" and
+// val can contain any char except ";~".
+//
+// The <value> is the textual representation of the metric value.
+//
+// The <timestamp> is the Unix time text of when the measurement was made.
+//
+// The returned values are:
+// 	- a string concatenating all generated "lines" (each single one representing
+// 	  a single Carbon metric.
+//  - number of time series successfully converted to carbon.
+// 	- number of time series that could not be converted to Carbon.
+func metricDataToPlaintext(md consumerdata.MetricsData) (string, int, int) {
+	if len(md.Metrics) == 0 {
+		return "", 0, 0
+	}
+
+	var sb strings.Builder
+	numTimeseriesDropped := 0
+	totalTimeseries := 0
+	for _, metric := range md.Metrics {
+		totalTimeseries++
+		descriptor := metric.MetricDescriptor
+		name := descriptor.GetName()
+		if name == "" {
+			numTimeseriesDropped += len(metric.Timeseries)
+			// TODO: observability for this, debug logging.
+			continue
+		}
+
+		tagKeys := buildSanitizedTagKeys(metric.MetricDescriptor.LabelKeys)
+
+		for _, ts := range metric.Timeseries {
+			if len(tagKeys) != len(ts.LabelValues) {
+				numTimeseriesDropped++
+				// TODO: observability with debug, something like the message below:
+				//	"inconsistent number of labelKeys(%d) and labelValues(%d) for metric %q",
+				//	len(tagKeys),
+				//	len(labelValues),
+				//	name)
+
+				// From this point on all code below is safe to assume that
+				// len(tagKeys) is equal to len(labelValues).
+				continue
+			}
+
+			for _, point := range ts.Points {
+				timestampStr := formatInt64(point.GetTimestamp().GetSeconds())
```

Comment:
Is only "seconds" portion of the timestamp used? Is that Carbon's precision?

Reply:
Yes, that is Carbon's precision.